src/rgw/rgw_lc.cc

```diff
@@ -1249,32 +1251,262 @@ class LCOpAction_Transition : public LCOpAction {
+    ttier = target_rule.tier_targets.find(storage_class);
+    if (ttier != target_rule.tier_targets.end()) {
+      tier = ttier->second;
+    }
+    return 0;
+  }
+
+  int delete_tier_obj(lc_op_ctx& oc, RGWLCCloudTierCtx& tier_ctx) {
+    int ret = 0;
+
+    /* If bucket is versioned, create delete_marker for current version
+     */
+    if (oc.bucket->versioned() && oc.o.is_current() && !oc.o.is_delete_marker()) {
+        ret = remove_expired_obj(oc.dpp, oc, false);
+        ldpp_dout(oc.dpp, 20) << "delete_tier_obj Object(key:" << oc.o.key << ") current & not delete_marker" << " versioned_epoch:  " << oc.o.versioned_epoch << "flags: " << oc.o.flags << dendl;
+    } else {
+        ret = remove_expired_obj(oc.dpp, oc, true);
+        ldpp_dout(oc.dpp, 20) << "delete_tier_obj Object(key:" << oc.o.key << ") not current " << "versioned_epoch:  " << oc.o.versioned_epoch << "flags: " << oc.o.flags << dendl;
+    }
+    return ret;
+  }
+
+  int update_tier_obj(lc_op_ctx& oc, RGWLCCloudTierCtx& tier_ctx) {
+
+    map<string, bufferlist> attrs;
+    int r = 0;
+
+    real_time read_mtime;
+
+    std::unique_ptr<rgw::sal::Object::ReadOp> read_op(oc.obj->get_read_op(&oc.rctx));
+
+    read_op->params.lastmod = &read_mtime;
+
+    r = read_op->prepare(null_yield, oc.dpp);
+    if (r < 0) {
+      return r;
+    }
+
+    if (read_mtime != tier_ctx.o.meta.mtime) {
+      /* raced */
+      return -ECANCELED;
+    }
+
+    attrs = oc.obj->get_attrs();
+    (*tier_ctx.obj)->set_atomic(&tier_ctx.rctx);
+    
+    RGWObjState *s = tier_ctx.rctx.get_state((*tier_ctx.obj)->get_obj());
+    std::unique_ptr<rgw::sal::Object::WriteOp> obj_op(oc.obj->get_write_op(&oc.rctx));
+
+    obj_op->params.modify_tail = true;
+    obj_op->params.flags = PUT_OBJ_CREATE;
+    obj_op->params.category = RGWObjCategory::CloudTiered;
+    obj_op->params.delete_at = real_time();
+    bufferlist blo;
+    blo.append("");
+    obj_op->params.data = &blo;
+    obj_op->params.if_match = NULL;
+    obj_op->params.if_nomatch = NULL;
+    obj_op->params.user_data = NULL;
+    obj_op->params.zones_trace = NULL;
+    obj_op->params.delete_at = real_time();
+    obj_op->params.olh_epoch = tier_ctx.o.versioned_epoch;
+    
+    RGWObjManifest *pmanifest; 
+
+    pmanifest = &(*s->manifest);
+    RGWObjTier tier_config;
+    tier_config.name = oc.tier.storage_class;
+    tier_config.tier_placement = oc.tier;
+    tier_config.is_multipart_upload = tier_ctx.is_multipart_upload;
+
+    pmanifest->set_tier_type("cloud-s3");
+    pmanifest->set_tier_config(tier_config);
+
+    /* check if its necessary */
+    rgw_placement_rule target_placement;
+    target_placement.inherit_from(tier_ctx.bucket_info.placement_rule);
+    target_placement.storage_class = oc.tier.storage_class;
+    pmanifest->set_head(target_placement, (*tier_ctx.obj)->get_obj(), 0);
+
+    pmanifest->set_tail_placement(target_placement, (*tier_ctx.obj)->get_obj().bucket);
```

Comment:
@cbodley  please let me know if the updates to object manifest done here are sufficient? Is there a way to retain head obj but delete only tail objects from manifest->objs?